phpBB/phpbb/captcha/plugins/honeypot.php

```diff
@@ -0,0 +1,358 @@
+	public function acp_page($id, &$module)
+	{
+	}
+
+	/**
+	 * ACP template to be displayed as demo
+	 *
+	 * @param int $id Module ID
+	 * @return string
+	 */
+	public function get_demo_template($id)
+	{
+		return 'captcha_honeypot.html';
+	}
+
+	/**
+	 * Prepare template for display
+	 *
+	 * @return string Template filename
+	 */
+	public function get_template()
+	{
+		$this->template->assign_vars([
+			'HONEYPOT_LABEL'	=> $this->honeypot_bank[$this->code],
+			'HONEYPOT_NAME'		=> $this->code,
+			'S_TYPE'			=> $this->type,
+			'HONEYPOT_CLASS'	=> self::CLASSES[array_rand(self::CLASSES)],
+		]);
+
+		return 'captcha_honeypot.html';
+	}
+
+	// not needed
+	public function execute()
+	{
+	}
+
+	/**
+	 * Get hidden fields for the current form
+	 *
+	 * @return array Hidden fields needed by this CAPTCHA
+	 */
+	public function get_hidden_fields()
+	{
+		return [
+			'confirm_id'	=> $this->confirm_id,
+		];
+	}
+
+	/**
+	 * Validate user's response
+	 *
+	 * @return string|bool Error string or false if response is OK
+	 */
+	public function validate()
+	{
+		if (!$this->user->is_setup())
+		{
+			$this->user->setup();
+		}
+
+		$error = '';
+		if (!$this->confirm_id)
+		{
+			$error = $this->language->lang('CONFIRM_CODE_WRONG');
+		}
+		else
+		{
+			if ($this->request->variable($this->code, '') === '')
+			{
+				$this->solved = true;
+			}
+			else
+			{
+				$error = $this->language->lang('CONFIRM_CODE_WRONG');
+			}
+		}
+
+		if (strlen($error))
+		{
+			// okay, incorrect answer. Let's ask a new question.
+			$this->new_attempt();
+			return $error;
+		}
+		else
+		{
+			return false;
+		}
+	}
+
+	/**
+	 * Is CAPTCHA correctly solved?
+	 *
+	 * @return boolean
+	 */
+	public function is_solved()
+	{
+		if ($this->solved === 0)
+		{
+			$this->validate();
+		}
+		return (bool) $this->solved;
+	}
+
+	/**
+	 * Get number of attempts
+	 *
+	 * @return int
+	 */
+	public function get_attempt_count()
+	{
+		return $this->attempts;
+	}
+
+	// not needed
+	public function install()
+	{
+	}
+
+	/**
+	 * Uninstall extension
+	 */
+	public function uninstall()
+	{
+		$this->garbage_collect(0);
+	}
+
+	/**
+	 * Generate new honeypot
+	 */
+	protected function generate_honeypot()
+	{
+		$this->confirm_id = md5(unique_id($this->user->ip));
+		$this->code = array_rand($this->honeypot_bank);
+		$this->solved = 0;
+
+		$sql = 'INSERT INTO ' . $this->confirm_table . ' ' . $this->db->sql_build_array('INSERT', [
+			'confirm_id'	=> (string) $this->confirm_id,
+			'session_id'	=> (string) $this->user->session_id,
+			'confirm_type'	=> (int) $this->type,
+			'code'			=> (string) $this->code,
+		]);
+		$this->db->sql_query($sql);
+	}
+
+	/**
+	 * Regenerate honeypot
+	 */
+	protected function new_attempt()
+	{
+		$this->code = array_rand($this->honeypot_bank);
+		$this->solved = 0;
+
+		$sql = 'UPDATE ' . $this->confirm_table . ' SET ' . $this->db->sql_build_array('UPDATE', [
+				'code'	=> (string) $this->code,
+			]) . "
+			, attempts = attempts + 1
```

Comment:
I'd put the comma at the end of the previous line and indent this one level more to make it a bit more readable.